### PR TITLE
Move dask integration to hdfs3 

### PR DESCRIPTION
**Pull request by jcrist**
Move the hdfs & dask integration out of `distributed.hdfs`. This removes
the need for managing hdfs infrastructure in the distributed tests, and
consolidates related code. Subsequent PRs will be made to distributed
and dask to cleanup references.

Also adds a pyarrow wrapper to support reading from hdfs using pyarrow.

### CODE REVIEW

**Comment by mrocklin** [comment]
I'm definitely in favor of moving hdfs testing code out of the dask/distributed repository.  It feels a little bit weird (though also very convenient) to move it into the dask/hdfs3 repository.  I'm fine either way.  

I would be curious to hear the perspective of @martindurant

**Comment by jcrist** [comment]
My thought was:
- `hdfs3` already has the testing infrastructure setup
- `hdfs3` implements a filesystem object
- dask integration is basically registering a filesystem object with dask, the actual code needed is fairly small.
- `gcsfs` does a similar thing, adding the dask integration as an optional submodule.

In the future once the interface has standardized a bit, and https://github.com/dask/dask/issues/2885 is fixed, the code needed for dask integration should be mostly trivial - ideally there'd be no need for a separate `DaskHDFileSystem` class.

**Comment by jcrist** [comment]
I should also note that except for a few cases (e.g. serialization works fine), there should be no real reason to test this with distributed - it works fine with the local schedulers. I just copied the tests over, but realistically most of them could use the threaded/serial scheduler instead.

**Comment by mrocklin** [comment]
Historically the reason why we had separate tests with the distributed scheduler was in order to test the use of worker restrictions for data locality.  Ideally dask workers prefer to read data from machines that have HDFS blocks locally.

**Comment by jcrist** [comment]
> Historically the reason why we had separate tests with the distributed scheduler was in order to test the use of worker restrictions for data locality. Ideally dask workers prefer to read data from machines that have HDFS blocks locally.

I'm unable to find either tests or code for this - can you provide a link? If the worker reading happens to be on the same node as their data then a shortcircuit read (managed by libhdfs3) may be permitted if shortcircuits are allowed in `hdfs-site.xml`. I see no code in any hdfs specific handling that would make use of finding blocks to direct the scheduler.

**Comment by mrocklin** [comment]
Search for `get_block_locations` in dask/bytes/core.py .  It's a bit hacky in that it imports dask.distributed at that point to see if we're in a distributed situation.

**Comment by jcrist** [comment]
Ah, I see it now. Thanks. So at least some of the tests do still need distributed.

File: hdfs3/dask.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+        else:
+            assert isinstance(host, str)
+            return host
+
+
+# Older versions of dask will try to load hdfs support from
+# distributed.hdfs. We check this import first to then override
+# it later.
+try:
+    import distributed.hdfs  # noqa
```
Comment (martindurant) [inline hdfs3/dask.py]: deprecation?

File: hdfs3/_pyarrow.py [hunk 2]
```diff
@@ -0,0 +1,12 @@
+from __future__ import absolute_import
+
+import pyarrow as pa
+
+
+class HDFSWrapper(pa.filesystem.DaskFileSystem):
```
Comment (martindurant) [inline hdfs3/_pyarrow.py]: Why is this required when isdir and isfile are already methods of DaskHDFileSystem? Can you derrive from both classes instead?

File: hdfs3/_pyarrow.py [hunk 3]
```diff
@@ -0,0 +1,12 @@
+from __future__ import absolute_import
+
+import pyarrow as pa
```
Comment (martindurant) [inline hdfs3/_pyarrow.py]: Does there need to be a deprecated import here too, like with distributed.hdfs, below?

File: hdfs3/tests/test_dask.py [hunk 4]
```diff
@@ -0,0 +1,461 @@
+from __future__ import print_function, division, absolute_import
+
+import json
+import sys
+import pytest
+from contextlib import contextmanager
+
+pytest.importorskip('distributed')
```
Comment (martindurant) [inline hdfs3/tests/test_dask.py]: Comment requested: we normally *do* perform these tests.

File: hdfs3/_pyarrow.py (hunk 3 above)
Comment (jcrist) [reply hdfs3/_pyarrow.py]: No, this is loaded by `dask.py` only. New functionality.

File: hdfs3/_pyarrow.py (hunk 2 above)
Comment (jcrist) [reply hdfs3/_pyarrow.py]: Pyarrow requires classes to be subclasses of its filesystem. This is how other pyarrow classes wrapper classes are handled. Subclassing from both might also work but wouldn't necesssarily be simpler. This class should also disappear on a soon-ish basis.

File: hdfs3/dask.py (hunk 1 above)
Comment (jcrist) [reply hdfs3/dask.py]: See the comment above these lines. This is to ensure that any other version of `DaskHDFilesystem` is registered first, so that it doesn't override this one. The new releases of dask/distributed will have this migrated, so this could be removed then.

File: hdfs3/tests/test_dask.py (hunk 4 above)
Comment (jcrist) [reply hdfs3/tests/test_dask.py]: Yes, they are run on travis. It's important that the test suite run fine with only required dependencies (or which dask/distributed are not).

**Comment by jcrist** [comment]
> also, I think this code no longer explicitly depends on distributed, only dask - the block locations are used only if there is a distributed client.

This is correct, but there are a few tests copied over that do explicitly check that this is used on distributed properly. The copied over tests could probably be rewritten to require distributed only for those tests, and just use the local schedulers for everything else, but I'd like to put that off for later if ever.

**Comment by martindurant** [comment]
+1
Further rewrites can wait for non-urgent follow-ups.
